### PR TITLE
HPCC-18607 Do not generate stats for edges leaving sink activities

### PR DESCRIPTION
**Pull request by shamser**
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by shamser** [comment]
@ghalliday

**Comment by ghalliday** [comment]
@shamser the change looks correct.  You modified the title of the PR, but not the commit.
Better would be
HPCC-18607 Do not generate stats for edges leaving sink activities

File: ecl/hthor/hthor.ipp [hunk 1]
```diff
@@ -188,7 +188,7 @@ static bool verifyFormatCrcSuper(unsigned helperCrc, IDistributedFile * df, bool
 }
 
 #define IMPLEMENT_SINKACTIVITY \
-    virtual unsigned queryOutputs() { return 0; } \
+    virtual unsigned queryOutputs() const override { return 0; } \
```
Comment (ghalliday) [inline ecl/hthor/hthor.ipp]: Adding override to just one function causes problems with clang (especially apple)

File: ecl/hthor/hthor.ipp [hunk 2]
```diff
@@ -188,7 +188,7 @@ static bool verifyFormatCrcSuper(unsigned helperCrc, IDistributedFile * df, bool
...
     virtual const void * nextRow() { throwUnexpected(); } \
```
Comment (ghalliday) [inline ecl/hthor/hthor.ipp]: these will also need override

**Comment by shamser** [comment]
@ghalliday Modified as suggested.  Please can your review

**Comment by ghalliday** [comment]
@shamser all the functions in the macro definition also need override attributes.

**Comment by shamser** [comment]
@ghalliday Hopefully, all good now?

**Comment by ghalliday** [comment]
@shamser I don't have an apple compiler to test this change on, but I don't think it is correct.  The warning is generated if some members of a class have override on them, but others do not.

That means that any class that uses IMPLEMENT_SINKACTIVITY inside its definition will also need override on all the other virtuals in that class.  There are at least 10 or more examples.  (You also need to change IMPLEMENT_IINTERFACE to IMPLEMENT_IINTERFACE_O in CHThorActivityBase.)

I suspect the best way to get this correct is to add it to all the classes in hthor.ipp.  However that should really be done as a separate change since it will swamp the original change.  So please can you remove all but the essential change, and then create a new PR for the override changes.

**Comment by shamser** [comment]
@ghalliday Would it be sensible to remove the override attribute for now?

**Comment by shamser** [comment]
@ghalliday Removed override attribute from function definition as discussed in flow.

**Comment by ghalliday** [comment]
@shamser looks good, please squash

**Comment by shamser** [comment]
@ghalliday Squashed

**Pull request merged**